### PR TITLE
Better Windows support

### LINKED ISSUE #8825: Zed main branch build windows crash
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Just clone main branch, run `cargo run`
It complains could not find conPTY, which is definetely there since I am using Windows11

### Environment

Windows 11 SImplified Chinese version

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

[2024-03-04T21:31:40+08:00 ERROR fs] watch path does not exist: C:\Users\36477\.config\zed\keymap.json
[2024-03-04T21:31:43+08:00 ERROR util] crates\zed\src/zed.rs:580: EOF while parsing a value at line 1 column 0
[2024-03-04T21:31:43+08:00 ERROR util] crates\zed\src/main.rs:718: 系统找不到指定的路径。 (os error 3)
[2024-03-04T21:31:43+08:00 ERROR fs] watch path does not exist: crates/zed/src/languages
[2024-03-04T21:31:43+08:00 ERROR fs] watch path does not exist: assets/icons/file_icons/file_types.json
[2024-03-04T21:31:44+08:00 ERROR util] crates\fs\src\repository.rs:101: Error { code: -9, klass: 4, message: "reference 'refs/heads/master' not found" }
[2024-03-04T21:31:46+08:00 ERROR fs] watch path does not exist: C:\Users\36477\.config\zed\tasks.json
Thread "main" panicked with "Unable to spawn shell: 文件名、目录名或卷标语法不正确。 (os error 123)" at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\alacritty_terminal-0.22.0\src\tty\windows\conpty.rs:235:5
   0: backtrace::backtrace::dbghelp::trace
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\backtrace-0.3.69\src\backtrace\dbghelp.rs:98
   1: backtrace::backtrace::trace_unsynchronized<backtrace::capture::impl$1::create::closure_env$0>
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\backtrace-0.3.69\src\backtrace\mod.rs:66
   2: backtrace::backtrace::trace<backtrace::capture::impl$1::create::closure_env$0>
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\backtrace-0.3.69\src\backtrace\mod.rs:53
   3: backtrace::capture::Backtrace::create
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\backtrace-0.3.69\src\capture.rs:176
   4: backtrace::capture::Backtrace::new
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\backtrace-0.3.69\src\capture.rs:140
   5: Zed::init_panic_hook::closure$0
             at C:\zjk\projects\zed\crates\zed\src\main.rs:622
   6: alloc::boxed::impl$49::call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\alloc\src\boxed.rs:2029
   7: std::panicking::rust_panic_with_hook
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:783
   8: std::panicking::begin_panic_handler::closure$0
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:657
   9: std::sys_common::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure_env$0,never$>
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\sys_common\backtrace.rs:171
  10: std::panicking::begin_panic_handler
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:645
  11: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\core\src\panicking.rs:72
  12: alacritty_terminal::tty::windows::conpty::panic_shell_spawn
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\alacritty_terminal-0.22.0\src\tty\windows\conpty.rs:235
  13: alacritty_terminal::tty::windows::conpty::new
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\alacritty_terminal-0.22.0\src\tty\windows\conpty.rs:220
  14: alacritty_terminal::tty::windows::new
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\alacritty_terminal-0.22.0\src\tty\windows\mod.rs:37
  15: terminal::TerminalBuilder::new
             at C:\zjk\projects\zed\crates\terminal\src\terminal.rs:385
  16: project::Project::create_terminal
             at C:\zjk\projects\zed\crates\project\src\terminals.rs:53
  17: terminal_view::impl$7::deserialize::closure$0::async_block$0::closure$1
             at C:\zjk\projects\zed\crates\terminal_view\src\terminal_view.rs:891
  18: gpui::window::impl$14::update_model<project::Project,enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,terminal_view::impl$7::deserialize::closure$0::async_block$0::closure_env$1>
             at C:\zjk\projects\zed\crates\gpui\src\window.rs:1692
  19: gpui::app::async_context::impl$3::update_model::closure$0<project::Project,enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,terminal_view::impl$7::deserialize::closure$0::async_block$0::closure_env$1>
             at C:\zjk\projects\zed\crates\gpui\src\app\async_context.rs:287
  20: gpui::app::impl$5::update_window::closure$0<enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,gpui::app::async_context::impl$3::update_model::closure_env$0<project::Project,enum2$<core::result::Result<gpui::app:
             at C:\zjk\projects\zed\crates\gpui\src\app.rs:1249
  21: gpui::app::AppContext::update<enum2$<core::result::Result<enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,anyhow::Error> >,gpui::app::impl$5::update_window::closure_env$0<enum2$<core::result::Result<gpui::app:
             at C:\zjk\projects\zed\crates\gpui\src\app.rs:360
  22: gpui::app::impl$5::update_window<enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,gpui::app::async_context::impl$3::update_model::closure_env$0<project::Project,enum2$<core::result::Result<gpui::app::entity_map
             at C:\zjk\projects\zed\crates\gpui\src\app.rs:1240
  23: gpui::app::async_context::impl$0::update_window<enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,gpui::app::async_context::impl$3::update_model::closure_env$0<project::Project,enum2$<core::result::Result<gpui::
             at C:\zjk\projects\zed\crates\gpui\src\app\async_context.rs:70
  24: gpui::app::async_context::impl$3::update_window<enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,gpui::app::async_context::impl$3::update_model::closure_env$0<project::Project,enum2$<core::result::Result<gpui::
             at C:\zjk\projects\zed\crates\gpui\src\app\async_context.rs:294
  25: gpui::window::AnyWindowHandle::update<gpui::app::async_context::AsyncWindowContext,enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,gpui::app::async_context::impl$3::update_model::closure_env$0<project::Project
             at C:\zjk\projects\zed\crates\gpui\src\window.rs:2633
  26: gpui::app::async_context::impl$3::update_model<project::Project,enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,terminal_view::impl$7::deserialize::closure$0::async_block$0::closure_env$1>
             at C:\zjk\projects\zed\crates\gpui\src\app\async_context.rs:286
  27: gpui::app::entity_map::Model::update<project::Project,gpui::app::async_context::AsyncWindowContext,enum2$<core::result::Result<gpui::app::entity_map::Model<terminal::Terminal>,anyhow::Error> >,terminal_view::impl$7::deserialize::closure$0::async_block$0::
             at C:\zjk\projects\zed\crates\gpui\src\app\entity_map.rs:406
  28: terminal_view::impl$7::deserialize::closure$0::async_block$0
             at C:\zjk\projects\zed\crates\terminal_view\src\terminal_view.rs:890
  29: core::future::future::impl$1::poll<alloc::boxed::Box<dyn$<core::future::future::Future<assoc$<Output,enum2$<core::result::Result<gpui::view::View<terminal_view::TerminalView>,anyhow::Error> > > > >,alloc::alloc::Global> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\future\future.rs:124
  30: async_task::runnable::impl$6::spawn_local::impl$1::poll<core::pin::Pin<alloc::boxed::Box<dyn$<core::future::future::Future<assoc$<Output,enum2$<core::result::Result<gpui::view::View<terminal_view::TerminalView>,anyhow::Error> > > > >,alloc::alloc::Global>
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\async-task-4.7.0\src\runnable.rs:455
  31: async_task::raw::RawTask::run<async_task::runnable::impl$6::spawn_local::Checked<core::pin::Pin<alloc::boxed::Box<dyn$<core::future::future::Future<assoc$<Output,enum2$<core::result::Result<gpui::view::View<terminal_view::TerminalView>,anyhow::Error> > >
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\async-task-4.7.0\src\raw.rs:557
  32: async_task::runnable::Runnable::run<tuple$<> >
             at C:\Users\36477\.cargo\registry\src\index.crates.io-6f17d22bba15001f\async-task-4.7.0\src\runnable.rs:781
  33: gpui::platform::windows::platform::impl$2::run
             at C:\zjk\projects\zed\crates\gpui\src\platform\windows\platform.rs:119
  34: gpui::app::App::run<Zed::main::closure_env$3>
             at C:\zjk\projects\zed\crates\gpui\src\app.rs:143
  35: Zed::main
             at C:\zjk\projects\zed\crates\zed\src\main.rs:120
  36: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\core\src\ops\function.rs:250
  37: std::sys_common::backtrace::__rust_begin_short_backtrace<void (*)(),tuple$<> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\std\src\sys_common\backtrace.rs:155
  38: std::sys_common::backtrace::__rust_begin_short_backtrace<void (*)(),tuple$<> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\std\src\sys_common\backtrace.rs:155
  39: std::rt::lang_start::closure$0<tuple$<> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\std\src\rt.rs:166
  40: std::rt::lang_start_internal::closure$2
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\rt.rs:148
  41: std::panicking::try::do_call
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:552
  42: std::panicking::try
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panicking.rs:516
  43: std::panic::catch_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\panic.rs:142
  44: std::rt::lang_start_internal
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library\std\src\rt.rs:148
  45: std::rt::lang_start<tuple$<> >
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce\library\std\src\rt.rs:165
  46: main
  47: invoke_main
             at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  48: __scrt_common_main_seh
             at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  49: BaseThreadInitThunk
  50: RtlUserThreadStart

error: process didn't exit successfully: `.\Zed.exe` (exit code: 0xffffffff)

### PR DESCRIPTION
**Pull request by JunkuiZhang**
- Menus, shortcuts, drag-drop, copy-paste....
- Using Blade and LinuxTextSystem to render
- Keyboards and mouse events are supported, IME is supported too
- With this [pr](https://github.com/alacritty/alacritty/pull/7796) merged, terminal will be easily supported

Edit:
Close #8825 

I will add some videos or pics below 

Release Notes:

- N/A

### CODE REVIEW

**Comment by JunkuiZhang** [comment]
屏幕截图 2024-03-04 145832
- menus
屏幕截图 2024-03-04 150613

- prompt
屏幕截图 2024-03-04 150448

- drag-drop

**Comment by JunkuiZhang** [comment]
What dont work:
1. System themes - The background color of menus wont change to black even if we set a dark theme
2. Scale factor - Currently, the scale factor is not functioning correctly. If a value greater than 1.0 is set, the right and bottom sides of the window will be pushed outside the window and become invisible. Since this value is consistently set to 1.0 in Linux, I suspect this issue may be related to Blade.
3. IME - The IME is functional, but the IME window position relies on caret position information. In other software like VS Code, the IME window always appears below the caret. However, Zed lacks caret position information, so currently the IME window will appear below the mouse cursor.

**Comment by JunkuiZhang** [comment]
## Dark theme but with white menus
屏幕截图 2024-03-03 003702
## Scale factor 1.5, note that Zed's "bottom bar" is off the screen
屏幕截图 2024-02-17 171720
## IME window position in Zed and VSCode
屏幕截图 2024-03-04 152112
屏幕截图 2024-03-04 152207

**Comment by JunkuiZhang** [comment]
As I am still in the learning process and do not have experience with large-scale projects, I am unsure if my code implementation follows best practices. Therefore, any suggestions and criticisms are welcome.

**Comment by zaucy** [comment]
hey looks like we both got started on better Windows support :) https://github.com/zed-industries/zed/pull/8791

I just gave your PR a shot and some things don't seem to work. Pressing 'esc' for instance just prints 'escape'. I've fixed those issues in my PR above, but you've added a bunch of things that I haven't gotten to yet. We should divide and conquer :)

Are you in the zed discord? If you aren't then you should join so we can all collaborate. https://zed.dev/docs/conversations

**Comment by JunkuiZhang** [comment]
> hey looks like we both got started on better Windows support :) #8791
> 
> I just gave your PR a shot and some things don't seem to work. Pressing 'esc' for instance just prints 'escape'. I've fixed those issues in my PR above, but you've added a bunch of things that I haven't gotten to yet. We should divide and conquer :)
> 
> Are you in the zed discord? If you aren't then you should join so we can all collaborate. https://zed.dev/docs/conversations

Currently, there are indeed many bugs, primarily requiring resolution of numerous merge conflicts for this PR, with some changes being rolled back. I have not yet joined Discord as the service is not available in China. I am considering ways to join.

**Comment by JunkuiZhang** [comment]
> hey looks like we both got started on better Windows support :) #8791
> 
> I just gave your PR a shot and some things don't seem to work. Pressing 'esc' for instance just prints 'escape'. I've fixed those issues in my PR above, but you've added a bunch of things that I haven't gotten to yet. We should divide and conquer :)
> 
> Are you in the zed discord? If you aren't then you should join so we can all collaborate. https://zed.dev/docs/conversations

Pressing `Esc` works fine here, that's stange. I will clean the build cache and test again.

**Comment by JunkuiZhang** [comment]
Currently, when attempting to open a folder in Zed, the application does not perform the acion. Upon checking the log, an error is reported stating that the folder cannot be found. Since it should be an absolute path, like "C:/xxx/xxx/xxx", but it gives the name directly, so I guess this may be a bug?
屏幕截图 2024-03-04 172723

**Comment by JunkuiZhang** [comment]
### Rust Analyzer works great

Note the buttom bar, it is running!!

屏幕截图 2024-03-04 233917

File: crates/gpui/build.rs [hunk 1]
```diff
@@ -21,6 +21,25 @@ fn main() {
     #[cfg(all(target_os = "macos", not(feature = "macos-blade")))]
     #[cfg(not(feature = "runtime_shaders"))]
     compile_metal_shaders(&header_path);
+
+    // We are using Windows Visita+, right..?
```
Comment (TDecking) [inline crates/gpui/build.rs:25]:
```suggestion
    // We are using Windows Vista+, right..?
```

Side note: The `rust-toolchain.toml` file mandates the Rust version of `1.76`, which is incidentally the first version to restrict official Windows support to [Windows 10 and Windows 11](https://github.com/rust-lang/compiler-team/issues/651).

Comment (JunkuiZhang) [reply crates/gpui/build.rs:25]: This actually is news to me. Additionally, the `alacritty_terminal` dependency that Zed relies on is implemented through conPTY API on Windows, while this API is only supported starting from the Windows 10 Fall 2018 update. So, I guess these users will also be left behind..?

**Comment by zaucy** [comment]
@JunkuiZhang this is what my experience was like running the latest commit on this PR `01cdc7038f58acb54445756c3bc22b674c9be3b7`

when I first do `cargo run` it's stuck in this state for a little bit
image

then finally I can resize it to get it visible it behaves like this:
pr-8809

sometimes when I click it creates multi cursors as well

**Comment by JunkuiZhang** [comment]
This is quite strange. It runs perfectly fine on my end. I will upload a video

**Comment by JunkuiZhang** [comment]
https://github.com/zed-industries/zed/assets/14981363/00fbbe09-7905-4582-bf4d-4cefd39a96b7

**Comment by JunkuiZhang** [comment]
I use win + G to record this, it dose not record the popups

**Comment by ThatFrogDev** [comment]
> I use win + G to record this, it dose not record the popups

@JunkuiZhang: if you are on the latest version of Windows 11 you can also use Win+Shift+S and select the area you want to record :)

**Comment by JunkuiZhang** [comment]
> > I use win + G to record this, it dose not record the popups
> 
> @JunkuiZhang: if you are on the latest version of Windows 11 you can also use Win+Shift+S and select the area you want to record :)

Thx for the tips!

**Comment by mikayla-maki** [comment]
> I have not yet joined Discord as the service is not available in China.

@JunkuiZhang I'm so sorry that y'all have to deal with that. Is there another place we could get in touch that's less asynchronous?

Regarding the two PRs, as CI is failing on this branch and https://github.com/zed-industries/zed/pull/8791 handles key events slightly better (e.g. the `esc` issue), I'm going to merge that one first.

But that said, I would really love to see the IME support in this PR land as well. It's very important that we're usable for everyone.

**Comment by mikayla-maki** [comment]
> 3\. but the IME window position relies on caret position information. In other software like VS Code, the IME window always appears below the caret. However, Zed lacks caret position information,

Zed is able to get this information from the `InputHandler::bounds_for_range` method :)

**Comment by mikayla-maki** [comment]
I understand, thank you for the work you've already put in!

**Pull request closed without merging**